### PR TITLE
Seurat 4 IO support and improve Seurat 3 IO support

### PR DESCRIPTION
**Pull request by pcm32**
This PR adds IO Support for Seurat 4 and improves IO support for Seurat 3 (Seurat >= 3.2 required possibly). Given that this package can be either installed with Seurat 3 or Seurat 4 (it just won't break provided that you don't call the wrong functions - clearly labeled). Tests are added for each case.

### CODE REVIEW

**Comment by pcm32** [comment]
I have tested this extensively both with Seurat 3 (https://github.com/ebi-gene-expression-group/r-seurat-scripts/pull/23) and Seurat 4 (https://github.com/ebi-gene-expression-group/r-seurat-scripts/pull/29), besides the added tests here. This is ready for review.

File: .github/workflows/pr-seurat-4.yaml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+          channels: defaults,conda-forge,bioconda
+          auto-activate-base: false
+          mamba-version: "*"
+          use-only-tar-bz2: true
+      - name: Tests
+        shell: bash -l {0}
+        env:
+          GITHUB_PAT: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          R -e 'remotes::install_github("satijalab/seurat-data", dependencies=FALSE)'
```
Comment (pcm32) [inline .github/workflows/pr-seurat-4.yaml:45]: Not sure why this is still here... but this needs to go away.

**Comment by pcm32** [comment]
Seurat 4 anndata reading won't work due to https://github.com/mojaveazure/seurat-disk/issues/30#issuecomment-914726678 - tried locally and this is still the case.

File: .github/workflows/pr-seurat-4.yaml (hunk 1 above)
Comment (pcm32) [reply .github/workflows/pr-seurat-4.yaml:45]: This is actually fine, I thought it was installing SeuratDisk, but Seurat Data is fine to install like this for tests.

**Comment by pcm32** [comment]
@pinin4fjords any objections on me merging this? I would then pin the current seurat-scripts in bioconda to the previous version and then do a release here, and then move seurat scripts to seurat 3.2.3 pointing to this new version. Thanks!

**Review by pinin4fjords — APPROVED** [review]
A lot of work here @pcm32 - nice! Looks good in principle, trust you on the detail.

**Comment by pcm32** [comment]
Tests passing for Seurat 3 and Seurat 4: https://github.com/ebi-gene-expression-group/r-seurat-scripts/pull/23 and https://github.com/ebi-gene-expression-group/r-seurat-scripts/pull/29, so will merge this.

**Pull request merged**